Add trade station cart

liach opened the pull request:
More leftovers:
Trade stations can dupe items as a result of the recent inv refactor. Needs fix.
Trade stations guis have weird color if f3+b bounding box render is on
The AI is weird, dunno if it really works.
The initial bytes for the trade station cart cannot really be improved given the data parameters are initialized before the fields in classes are. Decoding from each default logic is expensive and no less ugly.

**The Issue**
See #1093 comment for trade station cart.
 
**The Proposal**
 - Added carts module for disabling carts more easily
 - Changed how minecarts open guis (simpler now)
 - Moved trade station logic to a shared logic class for easy cleanups except entity/te specific parts
 - 
 
**Possible Side Effects**
 - Affects cart management in the future in config (new module)
 - Affects trade station behavior
 
**Alternatives**
No alternative is present except leaving this feature. The trade station cart reduce the need of moving villagers, which move on their own and might disrupt the minecarts' movements.

liach committed:
Add trade station cart

See #1093

Trade stations can dupe items as a result of the recent inv refactor. Needs fix.
Trade stations guis have weird color if f3+b bounding box render is on
The AI is weird, dunno if it really works.
The initial bytes for the trade station cart cannot really be improved given the data parameters are initialized before the fields in classes are. Decoding from each default logic is expensive and no less ugly.

Signed-off-by: liach <liach@users.noreply.github.com>

liach committed:
Make sure trading and ai setting only happens on the server; use fakeplayer

Signed-off-by: liach <liach@users.noreply.github.com>

liach committed:
Merge branch 'mc-1.12.2' into feature/trade-station-cart

liach commented:
the compile error is caused by the missing dependencies for the test source set.

CovertJaguar commented on src/main/java/mods/railcraft/client/gui/FactoryGui.java:
```diff
@@ -105,7 +106,7 @@
                     return new GuiFeedStation(inv, (TileFeedStation) obj);
                 case TRADE_STATION:
-                    return new GuiTradeStation(inv, (TileTradeStation) obj);
+                    return new GuiTradeStation(inv, obj instanceof TileTradeStation ? ((TileTradeStation) obj).getLogic() : ((EntityCartTradeStation) obj).getLogic(), (IWorldNameable) obj);
```
Use an interface?

CovertJaguar commented on src/main/java/mods/railcraft/common/gui/containers/FactoryContainer.java:
```diff
@@ -92,7 +93,7 @@ public static Container build(EnumGui gui, InventoryPlayer inv, Object obj, Worl
                     return new ContainerFeedStation(inv, (TileFeedStation) obj);
                 case TRADE_STATION:
-                    return new ContainerTradeStation(inv, (TileTradeStation) obj);
+                    return new ContainerTradeStation(inv, obj instanceof TileTradeStation ? ((TileTradeStation) obj).getLogic() : ((EntityCartTradeStation) obj).getLogic());
```
Again...interface

CovertJaguar commented on src/main/java/mods/railcraft/common/modules/ModuleExtras.java:
```diff
@@ -14,7 +14,7 @@
 import mods.railcraft.common.carts.RailcraftCarts;
 
-@RailcraftModule(value = "railcraft:extras", softDependencyClasses = ModuleTracks.class, description = "silly stuff, launch track, tnt carts, work cart")
+@RailcraftModule(value = "railcraft:extras", softDependencyClasses = {ModuleCarts.class, ModuleTracks.class}, description = "silly stuff, launch track, tnt carts, work cart")
```
Description needs an update.

CovertJaguar commented on src/main/java/mods/railcraft/common/util/entity/ai/EntityAIMoveToEntity.java:
```diff
@@ -0,0 +1,102 @@
+/**
+ * @author CovertJaguar <http://www.railcraft.info/>
+ */
+public class EntityAIMoveToEntity extends EntityAIBase {
```
Rename to EntityAISearchForEntity, and the block version to EntityAISearchForBlock.

CovertJaguar reviewed (changes requested):
Looks good generally, though the datawatcher stuff is definitely iffy.

liach committed:
Add container interface for logic so they are easier to use

Also addressed review concerns.

Todo: Refine network-related logics in the future! We will go and see.

Signed-off-by: liach <liach@users.noreply.github.com>

liach committed:
Merge branch 'feature/trade-station-cart' of https://github.com/Railcraft/Railcraft into feature/trade-station-cart

liach committed:
Merge branch 'mc-1.12.2' of https://github.com/Railcraft/Railcraft into feature/trade-station-cart

Signed-off-by: liach <liach@users.noreply.github.com>

liach commented:
wtf github removed that merge button here. stupid enough

CovertJaguar commented:
FYI, your merge failed.

>/home/travis/build/Railcraft/Railcraft/build/sources/main/java/mods/railcraft/common/util/chest/InventoryLogic.java:28: error: cannot find symbol
    public IInventory getInventory() {
           ^
  symbol:   class IInventory
  location: class InventoryLogic
/home/travis/build/Railcraft/Railcraft/build/sources/main/java/mods/railcraft/common/util/chest/TradeStationLogic.java:58: error: incompatible types: IInventoryComposite cannot be converted to IInventory
    private final InventoryAdvanced recipeSlots = new InventoryAdvanced(9).callbackInv(inventory).phantom();
                                                                                       ^
/home/travis/build/Railcraft/Railcraft/build/sources/main/java/mods/railcraft/common/util/chest/TradeStationLogic.java:64: error: incompatible types: IInventory cannot be converted to IInventoryComposite
        super(world, inventory);

liach commented:
yeah, been lazy to fix it.

liach committed:
Merge branch 'mc-1.12.2' into feature/trade-station-cart

CovertJaguar committed:
Merge branch 'mc-1.12.2' into feature/trade-station-cart

# Conflicts:
#	src/main/java/mods/railcraft/client/gui/TileGui.java

CovertJaguar committed:
Merge remote-tracking branch 'public/feature/trade-station-cart' into feature/trade-station-cart

CovertJaguar committed:
Rename tile classes to match naming scheme

CovertJaguar committed:
Remove a bunch of boilerplate and simplify logics in general.
